### DISCUSSION TITLE
[Feature Request] Twig filter to set the value of an array element or object property

### DISCUSSION THREAD

**Discussion by benface** [craftcms/cms#8826] [Ideas]
Inspired by the new `|push` and `|unshift` Twig filters in Craft 3.5, I would like to suggest a new filter to set the value of an array element (by key, numeric or string) or even of an object's public property. Here's an example to show you what I mean:

```twig
{% set page = {
  url: entry.url,
  title: entry.title,
} %}

{% set page = page|set('title', 'Other Title') %}
```

It would be nice if it worked with nested arrays/objects too:

```twig
{% set page = {
  url: entry.url,
  titles: {
    en: 'English Title',
    fr: 'French Title',
  },
} %}

{% set page = page|set('titles.en', 'Other English Title') %}
```

Or deeply nested:

```twig
{% set pages = [
  {
    url: entry1.url,
    titles: {
      en: 'Page 1 English Title',
      fr: 'Page 1 French Title',
    },
  },
  {
    url: entry2.url,
    titles: {
      en: 'Page 2 English Title',
      fr: 'Page 2 French Title',
    },
  },
] %}

{% set pages = pages|set('0.titles.en', 'Other English Title For Page 1') %}
```

If you're interested in adding this, I could make a PR for it.

**Comment by piotrpog** [comment]
It would be even better, if this was function/tag that worked on array directly, instead of just returning an updated array content.

**Comment by benface** [comment]
Interesting idea @piotrpog but I'm not sure I agree. I think it would be more consistent with everything else in Twig if it returned a new array. I think the filter syntax would be particularly misleading if not, although if it was a function it would be slightly less unexpected: `{% do set(array, 'key', 'value') %}`. But for objects, it's tricky... cloning an object just to change the value of a property can be expensive and rarely if ever what you want. And it would be super inconsistent if the behavior was different for arrays and objects. So maybe it shouldn't actually support PHP objects, just arrays and Twig "hashes" which are associative arrays.
